src/Language/Marlowe/Runtime/AdjustTip.hs

```diff
@@ -0,0 +1,92 @@
+module Language.Marlowe.Runtime.AdjustTip
+  ( BuildChain(..)
+  , adjustTip
+  , processTip
+
+  -- re-exporting
+  , Seq
+  , (><)
+  , empty
+  , fromList
+  , toList
+  )
+  where
+
+import Data.Foldable ( toList )  -- Used in this module as the counterpart to Sequence.fromList
+import Data.Monoid ( First(..), getFirst )
+import Data.Sequence ( Seq, (><), drop, elemIndexL, empty, fromList, mapWithIndex, take )
+import Prelude hiding ( drop, take )
+
+
+{-
+before
+
+0..9  10..19  20..24..29  30..34,<24>,35..39
+ t1     t2        t3             chain
+
+after
+
+0..24,35..39 
+  newChain
+
+So what are the cases?
+
+1. We found the divergence point, added new tip and used rest of old chain.
+   Chain is complete
+2. We didn't find divergence point, added new tip to new chain and must keep
+   looking. Chain is NOT COMPLETE
+3. We ran out of new tip segments to search for and they've all been added.
+   Chain is complete. (This represents the first time we run with an empty
+   chain. Like server start-up)
+
+1 and 3 seem like the same thing? (Just newChain)
+-}
+
+
+data BuildChain a
+  = Building (Seq a)  -- Didn't find divergence point yet, need more tip segments
+  | BuildDone (Seq a)  -- Found divergence point or old chain exhausted
+  deriving (Eq, Show)
+
+
+-- This function simulates the machinery of acquiring new tip segments and
+-- trying each one against the existing chain to compute a new chain in total.
+-- adjustTip isn't used by the Marlowe Explorer code but illustrates using
+-- processTip below
+adjustTip :: Eq a => Seq a -> [Seq a] -> Seq a
+adjustTip _ [] = empty
+adjustTip oldChain (tip : tips) = case processTip oldChain tip of
+  BuildDone endOfChain -> endOfChain
+  Building newTip -> newTip >< adjustTip oldChain tips
+
+
+-- Compute if and how a new tip maps onto an existing chain
+-- The state of this match is indicated through the BuildChain type
+processTip :: Eq a => Seq a -> Seq a -> BuildChain a
+processTip oldChain tip =
+  -- We use the First Monoid to isolate the first location where the new tip and
+  -- the old chain have a common element.
+  case getFirst . mconcat . map First . toList $ mapWithIndex (maybeMatch oldChain) tip of
```

Comment:
It would be simpler to use `breakl`, and return `fst` for `a` and a boolean `null . snd` for `Building` / `BuildDone`?